fix #297040: Palette keyboard navigation

shoogle opened the pull request:
Implements standard keyboard navigation for an item view / tree view as demonstrated by QAbstractItemView and QTreeView classes. Also changes how elements are displayed as necessary to indicate the different kinds of selection (e.g. when current item is not selected).

Resolves: https://musescore.org/en/node/297040

## Features

- Adds blue outline for current item (selected items have blue fill).
- Changing item also selects it (unless Ctrl is held).
- More button behaves like a normal item for navigation purposes

__New keyboard shortcuts:__

- Asterisk (*) key expands or collapses all items
- Menu key (or [Ctrl+]Shift+F10) opens context menus
- Press any letter key to jump to palette starting with that letter

__Various improvements to screen reader output:__

- Palette names end in "palette" to distinguish from elements
- Says when palettes are expanded
- Says how many elements in palette, and palettes in view
- Says when palette elements are selected

__Selection behavior__

Slightly changes behaviour for __mouse users__ to match that described in [QAbstractItemView::ExtendedSelection](https://doc.qt.io/qt-5/qabstractitemview.html#SelectionMode-enum):

> When the user selects an item in the usual way, the selection is cleared and the new item selected. However, if the user presses the Ctrl key when clicking on an item, the clicked item gets toggled and all other items are left untouched. If the user presses the Shift key while clicking on an item, all items between the current item and the clicked item are selected or unselected, depending on the state of the clicked item. Multiple items can be selected by dragging the mouse over them.

The behaviour for __keyboard users__ is almost the same, with point and click replaced by arrow keys and spacebar. Keyboard users also have the ability to change current item without changing the selection, which is done by holding Ctrl when using the arrow keys, as described [here](https://doc.qt.io/qt-5/qabstractitemview.html#details).

This behaviour for keyboard and mouse matches that of File Explorer in Windows 10.

## Checklist

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/musescore-coding-rules)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [N/A] I created the test (mtest, vtest, script test) to verify the changes I made

MarcSabatella commented:
Hmm, I accidentally click a checkbox in the original description now it shows as edited by me.

dmitrio95 commented:
Unfortunately with this PR palette cell background becomes white, hides palette grid and looks differently from the entire palette background. From the code I am not sure what is the reason of this but in master this seems to work correctly.
![Screenshot_20191225_095356](https://user-images.githubusercontent.com/6000747/71438056-cbbd0d00-26fc-11ea-8547-a66731105ff8.png)

shoogle commented on mscore/qml/palettes/Palette.qml:
```diff
@@ -387,36 +509,46 @@ GridView {
-
-                color: globalStyle.voice1Color
-                opacity: 0.0
+                color: "transparent"
```
@dmitrio95, fixed! I had to explicitly set the cell background color to "transparent" here, otherwise it was white by default hence the grid was no longer visible.

dmitrio95 commented:
This PR has some behavior conflicts with the latest changes implemented in #5546 as well as some visual issues. A brief list of issues I found while testing:

- Palette elements now apply on pressing a cell rather than clicking it. This makes dragging palette cell or its element cause unintended side effects if something is selected in a score.
- Palette cells do not change color when they are still pressed (or rather sometimes change). This is somewhat related to my previous point.
- Focused palette cell now has a blue frame around it (seems to be an intentional change, but what is this frame necessary for?)
- "Add to …" button in "More" popup now has a text saying something like "Add to Lines palette expanded"

Would it perhaps be better to merge only changes related to keyboard navigation first? Or are they tied to these visual and behavior changes as well?

shoogle commented:
@dmitrio95, I fixed the bugs. Sorry I missed them before. I rebased just before #5546 was merged and then had to rebase again immediately afterwards. I just did the minimum to get it working and didn't test properly.

The blue outline for the current item matches File Explorer in Windows is intentional behaviour. It is necessary in two situations.

### When the current item is not the only selected item:

![image](https://user-images.githubusercontent.com/11011881/71489854-ad9bfd80-281f-11ea-876b-e7ad35574464.png)

c.f. [File Explorer](https://user-images.githubusercontent.com/11011881/71489909-f8b61080-281f-11ea-92f8-502a6319504d.png)

### When the current item is not selected:

![image](https://user-images.githubusercontent.com/11011881/71489680-a9231500-281e-11ea-8957-674ad6b5e048.png)

c.f. [File Explorer](https://user-images.githubusercontent.com/11011881/71489925-184d3900-2820-11ea-9248-47dedf43c7bf.png)

In these situations it is necessary to show which item has keyboard focus using something other than selection colour.

> Would it perhaps be better to merge only changes related to keyboard navigation first? Or are they tied to these visual and behavior changes as well?

They are tied together. I think it should be safe to merge this as-is. The bugs so far have been cosmetic in nature rather that functional. If anything else comes up then ping me on Telegram and I can fix them easily enough.

Tantacrul commented:
Works very nicely, Peter. Just checked it out. Cheers.

MarcSabatella commented:
I should have commented about that also, I was using this code extensively during the run-up to FOSDEM (via the special FOSDEM PR I had created), and I agree, it works very nicely.  I still wish we could solve the problem of crashes when navigating the palette when a screenreader is active using Qt 5.9 (it's fixed in 5.12 for me) but it doesn't seem to be our fault, and in any case, this PR doesn't make it worse that I can see.
